### PR TITLE
[DataTable] Fixes bug in `DataTable` tooltips when using `truncate` prop

### PR DESCRIPTION
**Pull request by philschoefer**
### WHY are these changes introduced?

The current implementation of tooltips for overflowing column content in `DataTable` has a bug where tooltip content doesn't get updated properly when rows are re-sorted (e.g. using headers to sort the table).

<details>
<summary>Video of the issue</summary>

Tooltip becomes disassociated from content

</details>

This PR updates and isolates the logic.

@nickpresta @tylernoseworthy and I had discussed that there might be a benefit in having `TruncatedText` be it's own component at some point, but opted against it in this PR since this is addressing a production bug and that work would require a bit more thinking around what the API for this would ideally look like.

Since hovering over any cell triggers a re-render (to set the appropriate `-hovered` `className` on each cell) this also keeps checking whether each particular cell still overflows in case this changed (either due to new content, increased font size, viewport, etc.)

### How to 🎩

1. Set the Playground to the provided code
2. Hover over an overflowing cell
3. You should see the correct tooltip for cells that overflow and no tooltip for the cells that don't.
4. Sort by "Total Sales"
5. Hover over an element that previously had a tooltip and verify that it no longer does (in case it no longer overflows) or it does, but with the correct content (in case it does overflow).
6. Increase your browsers font-size and verify the tooltip shows correctly in cases where the font size changed triggered a cell to overflow when it previously wasn't.

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React, {useState} from 'react';

import {Link, Card, DataTable, Page} from '../src';

export function Playground() {
  const rows = [
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="emerald-silk-gown"
      >
        Emerald Silk Gown
      </Link>,
      '$875.00',
      124689,
      140,
      '$121,500.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="mauve-cashmere-scarf"
      >
        Mauve Cashmere Scarf
      </Link>,
      '$230.00',
      124533,
      83,
      '$19,090.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="navy-merino-wool"
      >
        Navy Merino Wool Blazer with khaki chinos and yellow belt
      </Link>,
      '$445.00',
      124518,
      32,
      '$14,240.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="emerald-silk-gown"
      >
        Emerald Silk Gown
      </Link>,
      '$875.00',
      124689,
      140,
      '$121,500.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="mauve-cashmere-scarf"
      >
        Mauve Cashmere Scarf
      </Link>,
      '$230.00',
      124533,
      83,
      '$19,090.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="navy-merino-wool"
      >
        Navy Merino Wool Blazer with khaki chinos and yellow belt
      </Link>,
      '$445.00',
      124518,
      32,
      '$14,240.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="emerald-silk-gown"
      >
        Emerald Silk Gown
      </Link>,
      '$875.00',
      124689,
      140,
      '$121,500.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="mauve-cashmere-scarf"
      >
        Mauve Cashmere Scarf
      </Link>,
      '$230.00',
      124533,
      83,
      '$19,090.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="navy-merino-wool"
      >
        Navy Merino Wool Blazer with khaki chinos and yellow belt
      </Link>,
      '$445.00',
      124518,
      32,
      '$14,240.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="emerald-silk-gown"
      >
        Emerald Silk Gown
      </Link>,
      '$875.00',
      124689,
      140,
      '$121,500.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="mauve-cashmere-scarf"
      >
        Mauve Cashmere Scarf
      </Link>,
      '$230.00',
      124533,
      83,
      '$19,090.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="navy-merino-wool"
      >
        Navy Merino Wool Blazer with khaki chinos and yellow belt
      </Link>,
      '$445.00',
      124518,
      32,
      '$14,240.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="emerald-silk-gown"
      >
        Emerald Silk Gown
      </Link>,
      '$875.00',
      124689,
      140,
      '$121,500.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="mauve-cashmere-scarf"
      >
        Mauve Cashmere Scarf
      </Link>,
      '$230.00',
      124533,
      83,
      '$19,090.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="navy-merino-wool"
      >
        Navy Merino Wool Blazer with khaki chinos and yellow belt
      </Link>,
      '$445.00',
      124518,
      32,
      '$14,240.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="emerald-silk-gown"
      >
        Emerald Silk Gown
      </Link>,
      '$875.00',
      124689,
      140,
      '$121,500.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="mauve-cashmere-scarf"
      >
        Mauve Cashmere Scarf
      </Link>,
      '$230.00',
      124533,
      83,
      '$19,090.00',
    ],
    [
      <Link
        removeUnderline
        url="https://www.example.com"
        key="navy-merino-wool"
      >
        Navy Merino Wool Blazer with khaki chinos and yellow belt
      </Link>,
      '$445.00',
      124518,
      32,
      '$14,240.00',
    ],
  ];
  const [sortedRows, setSortedRows] = useState<any[]>(rows);

  return (
    <Page title="Sales by product">
      <Card>
        <DataTable
          columnContentTypes={[
            'text',
            'numeric',
            'numeric',
            'numeric',
            'numeric',
          ]}
          headings={[
            'Product',
            'Price',
            'SKU Number',
            'Net quantity',
            'Net sales',
          ]}
          rows={sortedRows}
          totals={['', '', '', 255, '$155,830.00']}
          sortable={[true, true, false, false, true]}
          defaultSortDirection="descending"
          initialSortColumnIndex={4}
          truncate
          onSort={(index, direction) => {
            setSortedRows(
              [...rows].sort((rowA, rowB) => {
                const amountA = rowA[index]?.props?.children
                  ? rowA[index]?.props?.children
                  : parseFloat(rowA[index].substring(1));
                const amountB = rowB[index]?.props?.children
                  ? rowB[index]?.props?.children
                  : parseFloat(rowB[index].substring(1));

                return (
                  direction === 'descending'
                    ? amountA > amountB
                    : amountB > amountA
                )
                  ? 1
                  : -1;
              }),
            );
          }}
          footerContent={`Showing ${sortedRows.length} of ${sortedRows.length} results`}
          stickyHeader
          hasFixedFirstColumn
        />
      </Card>
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Comment by philschoefer** [comment]
Spin Instance running with this PR in `shopify/web`: https://shop1.shopify.polaris-update-9-23-1.philipp-schofer.us.spin.dev/admin/spin/tophat?activate=true

File: polaris-react/src/components/DataTable/components/Cell/Cell.tsx [hunk 1]
```diff
@@ -203,5 +182,37 @@ export function Cell({
...
+  // const handleCellFocus = (event: React.FocusEvent<HTMLElement>) => {
+  //   const anchor = containerRef?.current?.querySelector('a');
+
+  //   if (anchor === event.target) {
+  //     setIsFocused(true);
+  //   }
+
+  //   setShowTooltip(true);
+  // });
+
```
Comment (tylernoseworthy) [inline polaris-react/src/components/DataTable/components/Cell/Cell.tsx]: We can probably drop this code.

**Review by tylernoseworthy — APPROVED** [review]

**Review by nickpresta — APPROVED** [review]
It looks good in both Storybook and Web

File: .changeset/metal-impalas-type.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@shopify/polaris': patch
+---
+
+Fixes bug in `DataTable` where tooltips become disassociated when rows get sorted
```
Comment (laurkim) [inline .changeset/metal-impalas-type.md]:
```suggestion
Fixed bug in `DataTable` where tooltips become disassociated when rows get sorted
```

**Comment by philschoefer** [comment]
This PR contains a bug in Safari I'm currently investigating. Will re-request review once fixed.

**Comment by nickpresta** [comment]
@philschoefer is the bug in Safari caused by the changes in this PR? I still see it in the `main` branch: https://storybook.polaris.shopify.com/?path=/story/all-components-datatable--with-all-of-its-elements

If it's something easy to fix, it sounds good to bundle it in, but if it's a bigger thing, I think it's fine to do as a followup. What do you think?

**Comment by philschoefer** [comment]
I'll cut it off at 2pm today. I think I'm 99% there with just one thing pending.

I can add more details once I put up my changes.

**Comment by philschoefer** [comment]
I don't think the bug is caused by this PR. It's been there since the fixed first column implementation.

**Comment by philschoefer** [comment]
At heart, this is what the issue is: The `max-width` on the `th` is ignored by Safari but accepted by chrome. Overall this is more complicated to update because you have to consider how Safari's table rendering works and then apply `max-width` to elements within the cells instead all while keeping in mind elements having different padding.

Screen Shot 2022-08-11 at 1 52 52 PM

Since it's past the cut-off I'm gonna go ahead and merge this to deal with this bug later.

**Comment by philschoefer** [comment]
For context: The above bug only occurs when the font-size in the browser gets changed. By default everything works fine in Safari.

**Review by nickpresta — APPROVED** [review]

**Comment by nickpresta** [comment]
The Safari differences are quite frustrating :rage1: 

The 3 of us can pair on it next week if you want to work through it or bounce ideas around.

**Review by chloerice — APPROVED** [review]
🎩  LGTM 💯  :shipit:

**Pull request merged**